### ISSUE TITLE
windows version crashed

### ISSUE THREAD

**Issue by yangguanglu** [zeromq/libzmq#3054]
*Please use this template for reporting suspected bugs or requests for help.*

# Issue description
when run some time ,my program crashed .It seems like just happen on windows, Mac and linux are both ok.

# Environment

* libzmq version (commit hash if unreleased): 
* OS: 
4.2.1 and master
# Minimal test code / Steps to reproduce the issue

1.  

# What's the actual result? (include assertion message & call stack if applicable)

![wechatimg30](https://user-images.githubusercontent.com/8200462/39222804-1f68311c-4871-11e8-8874-114b746fd450.jpeg)

# What's the expected result?

**Comment by bluca** [comment]
Please provide code to reproduce the issue

**Comment by yangguanglu** [comment]
sorry I cannot find where it is.but I think the memory is alway increase.the call stack is this:

mvsd.exe!_invoke_watson(const wchar_t * expression, const wchar_t * function_name, const wchar_t * file_name, unsigned int line_number, unsigned __int64 reserved) 行 204 C++
  mvsd.exe!_invalid_parameter(const wchar_t * const expression, const wchar_t * const function_name, const wchar_t * const file_name, const unsigned int line_number, const unsigned __int64 reserved) 行 91 C++
  mvsd.exe!_invalid_parameter_noinfo() 行 98 C++
  mvsd.exe!__acrt_stdio_char_traits<char>::validate_stream_is_ansi_if_required(_iobuf * const stream) 行 433 C++
  mvsd.exe!__crt_stdio_output::output_processor<char,__crt_stdio_output::stream_output_adapter<char>,__crt_stdio_output::standard_base<char,__crt_stdio_output::stream_output_adapter<char> > >::process() 行 1608 C++
  mvsd.exe!common_vfprintf::__l11::<lambda>() 行 46 C++
  mvsd.exe!__crt_seh_guarded_call<int>::operator()<void <lambda>(void),int <lambda>(void) & __ptr64,void <lambda>(void) >(__acrt_lock_stream_and_call::__l3::void <lambda>(void) && setup, common_vfprintf::__l11::int <lambda>(void) & action, __acrt_lock_stream_and_call::__l4::void <lambda>(void) && cleanup) 行 199 C++
> mvsd.exe!__stdio_common_vfprintf(unsigned __int64 options, _iobuf * stream, const char * format, __crt_locale_pointers * locale, char * arglist) 行 59 C++
  mvsd.exe!fprintf(_iobuf * const _Stream, const char * const _Format, ...) 行 838 C++
  mvsd.exe!zmq::signaler_t::recv_failable() 行 352 C++
  mvsd.exe!zmq::mailbox_t::recv(zmq::command_t * cmd_, int timeout_) 行 89 C++
  mvsd.exe!zmq::reaper_t::in_event() 行 90 C++
  mvsd.exe!zmq::select_t::loop() 行 318 C++
  mvsd.exe!thread_routine(void * arg_) 行 47 C++
  mvsd.exe!thread_start<unsigned int (__cdecl*)(void * __ptr64)>(void * const parameter) 行 115 C++
  [外部代码]

**Comment by laplaceyang** [comment]
> but I think the memory is alway increase

Are you use two-way communication mode of zmq socket, but only recv never send / (only send never recv)?

For example:
1. use router but only send never recv or only recv never send.
2. use delear or other two-way communication mode and the usage is same as 1

I once used zmq socket described as example, and the memory always increased. And I solved it by solution descibed here [isuess2747](https://github.com/zeromq/libzmq/issues/2474)

**Comment by lxffff1990** [comment]
I think I found the problem. I found that there are some macros about assert in err.hpp in the libzmq library, include a lot of fprintf (stderr...
When I call fprintf in main function like this:
     for (int i = 0; i < 5; i++)
     {
         std::thread([]() {
             while(1)
                 fprintf(stderr, "connect error, %d/r/n", WSAGetLastError());
         }).detach();
     }

Dump occurs
This seems to be a fprintf duplicate  enter
How can I avoid this problem, or shield libzmq standard io output

My program runtime configuration is multithreaded debugging (/MTd), the same with libzmq.

**Comment by yangguanglu** [comment]
@laplaceyang thank you for your solution.can you answer the one upstairs~

**Comment by lxffff1990** [comment]
@sigiesec I debugged the source code, crashed in the recv_failable():379 line, the source code is:
<img width="335" alt="image" src="https://user-images.githubusercontent.com/19180081/39390861-a1fd3560-4acd-11e8-94fb-478d5404c763.png">
it is the MACRO wsa_assert
<img width="397" alt="image" src="https://user-images.githubusercontent.com/19180081/39390881-dcf273d8-4acd-11e8-98a0-b3dc95c40a96.png">
When recv failed, and error_code != WSAEWOULDBLOCK,it will write an error message to stderr
The crash occurs if I print an error message to stderr in my program at the same time

I wrote some test code, open five threads and write to stderr without crashing in main thread,its run ok
But when libzmq  write stderr at the same time  will crash

**Comment by lxffff1990** [comment]
@sigiesec Hello, I found another problem, in my program,I set stdout stdin,stderr encoding like:
#ifdef _MSC_VER
    if (_setmode(_fileno(file), _O_U8TEXT) == -1)
        throw std::runtime_error("Could not set STDIO to utf8 mode.");
#endif
Then it may be gbk by default in libzmq .
The different iostream encoding in different libraries may cause this problem.
Actually, I don't want other libraries to output to my console.

**Comment by sigiesec** [comment]
@lxffff1990 
Please do not post screenshots of the code, copy the code or even better link to the code line in the repository such as 
https://github.com/zeromq/libzmq/blob/4726f7262d076ed0b20111cfcdc0df6b33cf7c82/src/signaler.cpp#L379

In normal operation, libzmq will not write to stderr. As I wrote before, it will only do so if it encounters an internal error and calls abort immediately afterwards. Since it only writes ASCII, this will not even be a problem if you set output to UTF-8 mode. Apart from that, fprintf is thread-safe: https://stackoverflow.com/questions/22539282/is-msvcrts-implementation-of-fprintf-thread-safe

Something is wrong. This might be a bug in libzmq, but it is more likely that the cause is that your program uses a ZeroMQ socket from multiple threads, which is not allowed, and causes undefined behaviour.

**Comment by yangguanglu** [comment]
@sigiesec I agree, thank you!
